File: scripts/capture_report_screenshots.mjs
/**
 * Capture Myknowledge GUI screenshots for project report.
 * Prerequisite: backend running (python -m backend).
 *
 * Usage:
 *   cd e:\app\Myknowledge
 *   npx playwright install chromium
 *   node scripts/capture_report_screenshots.mjs
 */
import { chromium } from "playwright";
import { mkdir } from "fs/promises";
import path from "path";
import { fileURLToPath } from "url";

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
const OUT = path.join(ROOT, "docs", "report-screenshots");
const INDEX = path.join(ROOT, "electron", "renderer", "index.html");
const API = process.env.MYKNOWLEDGE_API || "http://127.0.0.1:18765";

const TABS = [
  ["ask", "01-tab-ask"],
  ["query", "02-tab-query"],
  ["list", "03-tab-notes"],
  ["produce", "04-tab-produce"],
  ["manage", "05-tab-import"],
  ["library", "06-tab-library"],
  ["history", "07-tab-history"],
];

async function waitBackend(page) {
  for (let i = 0; i < 30; i++) {
    try {
      const res = await page.request.get(`${API}/api/health`);
      if (res.ok()) return;
    } catch {
      /* retry */
    }
    await page.waitForTimeout(500);
  }
  throw new Error(`Backend not reachable at ${API}`);
}

async function shot(page, name) {
  const file = path.join(OUT, `${name}.png`);
  await page.screenshot({ path: file, fullPage: false });
  console.log("saved", file);
}

async function main() {
  await mkdir(OUT, { recursive: true });

  const browser = await chromium.launch({
    headless: true,
    args: ["--disable-web-security", "--allow-file-access-from-files"],
  });
  const context = await browser.newContext({
    viewport: { width: 1440, height: 900 },
    deviceScaleFactor: 1,
    colorScheme: "light",
  });
  await context.addInitScript((apiBase) => {
    window.myknowledge = {
      apiBase,
      deviceId: "report-screenshot",
      ensureFileViewer: async () => {},
      openAssetViewer: async () => {},
      pickDirectory: async () => null,
    };
  }, API);

  const page = await context.newPage();
  await waitBackend(page);

  const indexUrl = `file:///${INDEX.replace(/\\/g, "/")}`;
  await page.goto(indexUrl, { waitUntil: "networkidle", timeout: 60000 });
  await page.waitForSelector(".tab.active", { timeout: 15000 });
  await page.waitForTimeout(1200);

  await shot(page, "00-main-ask");

  for (const [tab, name] of TABS) {
    await page.click(`.tab[data-tab="${tab}"]`);
    await page.waitForSelector(`#panel-${tab}.active`, { timeout: 10000 });
    await page.waitForTimeout(800);
    if (tab === "library") {
      await page.waitForTimeout(1200);
    }
    await shot(page, name);
  }

  await page.click("#btn-settings");
  await page.waitForSelector("#settings-modal:not(.hidden)", { timeout: 10000 });
  await page.waitForTimeout(600);
  await shot(page, "08-modal-settings");

  await page.click("#btn-settings-cancel");
  await page.waitForTimeout(500);

  await page.click("#btn-persona");
  await page.waitForSelector("#persona-modal:not(.hidden)", { timeout: 10000 });
  await page.waitForTimeout(600);
  await shot(page, "09-modal-persona");

  await page.click("#btn-persona-cancel");
  await page.waitForSelector("#persona-modal.hidden", { timeout: 5000 });

  await browser.close();
  console.log("Done. Output:", OUT);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: scripts/capture_report_screenshots.py
"""Capture Myknowledge GUI screenshots for project report (Playwright).

Prerequisite: backend running — python -m backend

Usage:
    cd e:\\app\\Myknowledge
    python scripts/capture_report_screenshots.py
"""

from __future__ import annotations

import sys
import time
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
OUT = ROOT / "docs" / "report-screenshots"
INDEX = ROOT / "electron" / "renderer" / "index.html"
API = "http://127.0.0.1:18765"

TABS = [
    ("ask", "01-tab-ask"),
    ("query", "02-tab-query"),
    ("list", "03-tab-notes"),
    ("produce", "04-tab-produce"),
    ("manage", "05-tab-import"),
    ("library", "06-tab-library"),
    ("history", "07-tab-history"),
]


def wait_backend(page, timeout_s: float = 20.0) -> None:
    deadline = time.time() + timeout_s
    while time.time() < deadline:
        try:
            resp = page.request.get(f"{API}/api/health")
            if resp.ok:
                return
        except Exception:
            pass
        time.sleep(0.5)
    raise RuntimeError(f"Backend not reachable at {API}")


def main() -> int:
    try:
        from playwright.sync_api import sync_playwright
    except ImportError as e:
        print("Install: pip install playwright && python -m playwright install chromium", file=sys.stderr)
        raise SystemExit(1) from e

    OUT.mkdir(parents=True, exist_ok=True)
    index_url = INDEX.as_uri()

    with sync_playwright() as p:
        browser = p.chromium.launch(
            headless=True,
            args=["--disable-web-security", "--allow-file-access-from-files"],
        )
        context = browser.new_context(viewport={"width": 1440, "height": 900}, color_scheme="light")
        context.add_init_script(
            f'window.myknowledge = {{ apiBase: "{API}", deviceId: "report", '
            "ensureFileViewer: async () => {}, openAssetViewer: async () => {}, pickDirectory: async () => null };"
        )
        page = context.new_page()
        wait_backend(page)
        page.goto(index_url, wait_until="networkidle", timeout=60000)
        page.wait_for_selector(".tab.active", timeout=15000)
        page.wait_for_timeout(1200)

        def shot(name: str) -> None:
            path = OUT / f"{name}.png"
            page.screenshot(path=str(path), full_page=False)
            print("saved", path)

        shot("00-main-ask")
        for tab, name in TABS:
            page.click(f'.tab[data-tab="{tab}"]')
            page.wait_for_selector(f"#panel-{tab}.active", timeout=10000)
            page.wait_for_timeout(1000 if tab == "library" else 700)
            shot(name)

        page.click("#btn-settings")
        page.wait_for_selector("#settings-modal:not(.hidden)", timeout=10000)
        page.wait_for_timeout(600)
        shot("08-modal-settings")
        page.click("#btn-settings-cancel")
        page.wait_for_timeout(500)

        page.click("#btn-persona")
        page.wait_for_selector("#persona-modal:not(.hidden)", timeout=10000)
        page.wait_for_timeout(600)
        shot("09-modal-persona")

        browser.close()

    print("Done:", OUT)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
